### PR TITLE
Preventing crashing incase of no network by loading from cache

### LINKED ISSUE #91568: torch.hub.load() fails to load from cache if there is no network and crashes
### 🐛 torch.hub.load() fails to load from cache if there is no network and crashes

When trying to load a model from github, torch.hub.load() in pytorch/torch.hub.py fails to load from cache if there is no network. Since _parse_repo_info() does not handle network timeout error. 
Steps to reproduce:
* Please turn off internet
* Try the following code

```python
import torch
torch.hub.load("intel-isl/MiDaS", "transforms")
```

You may get the following exception

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/urllib/request.py", line 1348, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/lib/python3.10/http/client.py", line 1282, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.10/http/client.py", line 1328, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.10/http/client.py", line 1277, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.10/http/client.py", line 1037, in _send_output
    self.send(msg)
  File "/usr/lib/python3.10/http/client.py", line 975, in send
    self.connect()
  File "/usr/lib/python3.10/http/client.py", line 1447, in connect
    super().connect()
  File "/usr/lib/python3.10/http/client.py", line 941, in connect
    self.sock = self._create_connection(
  File "/usr/lib/python3.10/socket.py", line 824, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/lib/python3.10/socket.py", line 955, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/abuobaidazishan/Dev/DevProjects/autonom/test.py", line 2, in <module>
    torch.hub.load("intel-isl/MiDaS", "transforms")
  File "/home/abuobaidazishan/.local/lib/python3.10/site-packages/torch/hub.py", line 539, in load
    repo_or_dir = _get_cache_or_reload(repo_or_dir, force_reload, trust_repo, "load",
  File "/home/abuobaidazishan/.local/lib/python3.10/site-packages/torch/hub.py", line 180, in _get_cache_or_reload
    repo_owner, repo_name, ref = _parse_repo_info(github)
  File "/home/abuobaidazishan/.local/lib/python3.10/site-packages/torch/hub.py", line 134, in _parse_repo_info
    with urlopen(f"https://github.com/{repo_owner}/{repo_name}/tree/main/"):
  File "/usr/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.10/urllib/request.py", line 519, in open
    response = self._open(req, data)
  File "/usr/lib/python3.10/urllib/request.py", line 536, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 1391, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/usr/lib/python3.10/urllib/request.py", line 1351, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno -2] Name or service not known>

```
The used [github repo and branch](https://github.com/isl-org/MiDaS) is an example. Please, test with another repo, branch pair if needed.

### Versions

```
Collecting environment information...
PyTorch version: 1.13.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Pop!_OS 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-6.0.12-76060006-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.13.0+cpu
[pip3] torchaudio==0.13.0+cpu
[pip3] torchvision==0.14.0+cpu
[conda] Could not collect

```

cc @nairbv @NicolasHug @vmoens @jdsgomes

### PR DESCRIPTION
**Pull request by MohammedZ666**
Fixes #91568

### CODE REVIEW

File: torch/hub.py [hunk 1]
```diff
@@ -136,6 +136,16 @@ def _parse_repo_info(github):
                 ref = 'master'
             else:
                 raise
+        except OSError as e:  # Accounting for timeout error such that the code loads from cache when there is no internet
```
Comment (NicolasHug) [inline torch/hub.py:139]: Should we catch a more specific `URLError` instead of `OSError`?

File: torch/hub.py [hunk 2]
```diff
@@ -136,6 +136,16 @@ def _parse_repo_info(github):
...
+            ref = None
+            for possible_ref in ['main', 'master']:
+                is_ref_in_cache = os.path.exists(
+                    f"{get_dir()}/{repo_owner}_{repo_name}_{possible_ref}")
+                if is_ref_in_cache:
+                    ref = possible_ref
+                    break
+            if ref == None:
+                raise
```
Comment (NicolasHug) [inline torch/hub.py:148]: Code looks good, although I'd suggest to fail with an informative error message if the `ref` can't be found in the cache (also simplifying a few unnecessary variables):

```suggestion
        except URLError as e:
            # No internet connection, need to check for cache as last resort
            for possible_ref in ("main", "master"):
                if os.path.exists(f"{get_dir()}/{repo_owner}_{repo_name}_{possible_ref}"):
                    ref = possible_ref
                    break
            if ref is None:
                raise RuntimeError(
                    "It looks like there is no internet connection and the "
                    f"repo could not be found in the cache ({get_dir()})"
                ) from e
 
```

**Review by NicolasHug — COMMENTED** [review]
Thanks a lot for the report and for the fix @MohammedZ666 , the fix looks good. I made a suggestion below to simplify it and provide an error message, LMK what you think.

File: torch/hub.py (hunk 2 above)
Comment (MohammedZ666) [reply torch/hub.py:148]: Yes, you are right. However, I had to import URLError, maybe include that in your suggestion as well? **Line 15** of `pytorch:master/torch/hub.py` should be  of  should be: 

```
from urllib.error import HTTPError, URLError
```
Thanks

File: torch/hub.py (hunk 1 above)
Comment (MohammedZ666) [reply torch/hub.py:139]: Yes, that would be perfect.

File: torch/hub.py (hunk 2 above)
Comment (MohammedZ666) [reply torch/hub.py]: Committed above suggestions, please review.

**Comment by MohammedZ666** [comment]
@NicolasHug Sorry to bother you like this but the branch is still not merged, can you check?

**Review by NicolasHug — APPROVED** [review]
Thanks @MohammedZ666 , LGTM.

**Pull request closed without merging**